src/client/gfs_client.h

```diff
@@ -0,0 +1,44 @@
+#ifndef CLIENT_GFS_CLIENT_H
+#define CLIENT_GFS_CLIENT_H
+
+#include "google/protobuf/stubs/status.h"
+#include "google/protobuf/stubs/statusor.h"
+
+namespace gfs {
+namespace client {
+
+// Use the Status and Statusor objects provided by protobuf
+using google::protobuf::util::Status;
+using google::protobuf::util::StatusOr;
+
+/* Data type for the return of read operation. This includes how many bytes have been
+ * read and a pointer of the buffer */
+struct Data {
+   int bytesRead;
+   void* buffer; 
+};
+
+/* Provide an interface function to let user to initialize client and configure
+ * parameters such as master's url. A user should call this function before
+ * performing any of the following functions as initializations are needed */
+void configClient(const char* masterUrl);
+
+Status create(const char* pathname);
+Status remove(const char* pathname); 
+
+/* Let's plan to support the following open mode:
+ * Read Mode | Write Mode | Create Mode | Is Directory
+ * Possible combination: 
```

Comment:
2 points.
1. we don't need to have a directory mode right? GFS doesn't distinguish it explicitly 
2. how do we achieve combination modes, if we only have an int flags?

Reply:
Well, I was thinking about simplifying this too, but wouldn't it be troublesome if we don't distinguish a file from a directory, but then define a nested file in that file namespace? For example, if you have a file /foo, then it doesn't make sense to have another file /foo/bar. 

I went back and read the GFS paper again, and I think there is a distinction between file and directory (as there should be). Quotes from the paper "Typically, if it involves /d1/d2/.../dn/leaf, it will acquire read-locks on the directory names /d1, /d1/d2, ...,/d1/d2/.../dn, and either a read lock or a write lock on the full pathname /d1/d2/.../dn/leaf. Note that leaf may be a file or directory depending on the operation."

I can understand that they don't have an inode structure for directory, and this allows multiple creation to take place in a directory, but we have to distinguish a directory from a file, otherwise our system will look strange. 

The simplest way to support combination modes is to do so in the same way that open() system call deals with it. You can do this by bit mask. We don't have very complex combination so I am fine with either way. Mainly we just have to take care of create + write, and create + directory. I'd think using bit mask is a bit simpler.

Reply:
@xicheng87 
Based on everyday use of GFS, it’s like this, given a path:
`/d1/d2/d3/d4`, you read lock `/d1/d2/d3/`  and write lock `d4`. here, we don’t know if `d4` is a directory or file, but it’s fine because GFS doesn’t care!

If `d4` happens to be a directory and we need to concurrently write `/d1/d2/d3/d4/x` and `/d1/d2/d3/d4/y`, they will both read lock `/d1/d2/d3/d4` which is not blocking each other.

Reply:
Well, what if someone tries to write some stuff to /d1/d2/d3/d4 in your example? Should GFS care or not? 

In a sense, we are debating whether GFS should treat a directory in a traditional sense that it is purely a container for files and itself does not carry any data content. I don't think this has been clearly specified in the GFS paper. But you might have just given some insider's view. 

I think allowing a directory to carry data is fine, this directory can be seen as some kind of megafile where itself can have data and it can have child. So it's like a parent-child relationship with the data it holds, logically. 

I can see that this can simplify the design and impl and I think we should do this.

Reply:
Sounds good. FYI, yes, if someone tries to write on the directory `/d1/d2`, no one can write on the files underneath it `/d1/d2/x`. 

And yes, in GFS, based on my usecase, the directory actually is treated as a file and contains data. Think about this. Before, the GFS v1 doesn't support traditional file operations. Now, in unpublished GFS, they support normal directory operations such as `ls` and `chmod`, so the directory file now holds metadata.